src/AspNetCore/Src/applicationmanager.cxx

```diff
@@ -189,80 +191,250 @@ APPLICATION_MANAGER::GetApplicationInfo(
     {
-        // Application got recycled. Log an event
-        STACK_STRU(strEventMsg, 256);
-        if (SUCCEEDED(strEventMsg.SafeSnwprintf(
-            ASPNETCORE_EVENT_RECYCLE_CONFIGURATION_MSG,
-            pszApplicationId)))
+        if (m_hostingModel == HOSTING_IN_PROCESS)
         {
-            UTILITY::LogEvent(g_hEventLog,
-                EVENTLOG_INFORMATION_TYPE,
-                ASPNETCORE_EVENT_RECYCLE_CONFIGURATION,
-                strEventMsg.QueryStr());
+            // When we are inprocess, we need to keep the application the
+            // application manager that is being deleted. This is because we will always need to recycle the worker
+            // process and any requests that hit this worker process must be rejected (while out of process can
+            // start a new dotnet process). We will immediately call Recycle after this call.
+
+            // We need to clear the hash table here and the ref copy the original table to avoid memory leaks
+            // and keep the application referenced.
+            m_pApplicationInfoHash->Clear();
+            table->Apply(APPLICATION_INFO_HASH::ReferenceCopyToTable, static_cast<PVOID>(m_pApplicationInfoHash));
         }
-
     }
 
     if (m_pApplicationInfoHash->Count() == 0)
     {
         m_hostingModel = HOSTING_UNKNOWN;
     }
 
+    ReleaseSRWLockExclusive(&m_srwLock);
 
-    if (g_fAspnetcoreRHLoadedError)
+    if (!context.MultiSz.IsEmpty())
     {
-        // We had assembly loading failure
-        // this error blocked the start of all applications
-        // Let's recycle the worker process if user redeployed any application
-        g_pHttpServer->RecycleProcess(L"AspNetCore Recycle Process on Demand due to assembly loading failure");
-    }
+        PCWSTR path = context.MultiSz.First();
+        while (path != NULL)
+        {
+            APPLICATION_INFO* pRecord;
 
-    ReleaseSRWLockExclusive(&m_srwLock);
+            // Application got recycled. Log an event
+            STACK_STRU(strEventMsg, 256);
+            if (SUCCEEDED(strEventMsg.SafeSnwprintf(
+                ASPNETCORE_EVENT_RECYCLE_CONFIGURATION_MSG,
+                path)))
+            {
+                UTILITY::LogEvent(g_hEventLog,
+                    EVENTLOG_INFORMATION_TYPE,
+                    ASPNETCORE_EVENT_RECYCLE_CONFIGURATION,
+                    strEventMsg.QueryStr());
+            }
+
+            hr = key.Initialize(path);
+            if (FAILED(hr))
+            {
+                goto Finished;
+            }
+
+            table->FindKey(&key, &pRecord);
+            DBG_ASSERT(pRecord != NULL);
+
+            // shutdown will be done in another thread
+            RecycleApplication(pRecord, NULL);
+            pRecord->DereferenceApplicationInfo();
+            path = context.MultiSz.Next(path);
+        }
+    }
 
 Finished:
+    if (table != NULL && m_hostingModel != HOSTING_IN_PROCESS)
```

Comment:
only need to check table != NULL as already duplicated the entry back to m_pApplicationHaskTable. Otherwise, memory leak for the table

Reply:
Modifying to do a pointer copy instead of a table copy as I believe it will still work.